### PR TITLE
Unit tests for Notebook Instance

### PR DESCRIPTION
**Pull request by ananth102**
#82 is required for this PR.
This pull request has unit tests for the Notebook Instance, currently it has about 75% coverage and it takes about 0.1 seconds to run.

Here are all the tests:
Create:
- Create=InvalidInput - Given one of the parameters is invalid, ko.Status shows a terminal condition
- Create=Valid - Testing a successful create call

ReadOne:
- ReadOne=AfterCreate - Testing readOne right after create, the status should be in Pending with Resource synced being false.
- ReadOne=Deleting - Testing readOne when deleting, resource synced should be false.
- ReadOne=Fail - Testing readone when Describe fails, an error should be returned.
- ReadOne=FailedState - Testing how readone handles the Failed State, it should have a failure reason.
- ReadOne=InService - Testing how readone handles the InService state. Resource Synced should be true.
- ReadOne=StoppedAfterUpdateACK - Testing how readone handles the Stopped state after it's been updated. In this test the Notebook Instance has been stopped by the controller.
- ReadOne=StoppedAfterUpdateHuman - Testing how readone handles the Stopped state when stopped by a human for an update, StoppedByControllerMeta should not be in the status.
- ReadOne=StoppedBeforeUpdateACK - Testing how readone handles the Stopped state when it's stopped right before an update. The StoppedByControllerMetadata status should be in UpdatePending.
- ReadOne=StoppingDeleting - Testing how readone handles the Stopping state before going into deleting. The Resource synced condition should be set to False.
- ReadOne=StoppingUpdate - Testing how readone handles the Stopping state before going into Update. The resource synced condition should be False and the StoppedByControllerMetadata status should be set to UpdatePending.
- ReadOne=UpdatingUser - Testing how readone handles the Updating status when a user stops the Notebook. StoppedByControllerMetadata should not be in the status field.
- ReadOne=UpdatingUserStopping- Testing if the readone call updates the Notebook's status from stopped to updating, when it recieves information from the server.
- ReadOne=StartFail - Testing how readone would behave if start failed, test should return an error.

Update:
- Update=Pending - This test checks if the Notebook requeues while in Pending state.
- Update=Stopping - This test checks if the Notebook requeues while in Stopping state.
- Update=Updating - This test checks if the Notebook requeues while in updating state.
- Update=InService - This test checks if the controller stops the Notebook. Note: Although a stop controller call has been made runtime has not performed an sdk.Find() as a result the NotebookInstanceStatus will be InService.
- Update=Stopped - ByControllerUser - This test checks if the controller updates and does not set the StoppedByControllerMetadata status.
- Update=InvalidInput - This test checks what the controller would do if update failed. It should return an error.
- Update=Dissassociated - This test checks if the dissassociated parameters are being handled right.
- Update=StopError - This test checks if the controller returns an error if StopNotebookInstance fails.

Delete:
- Delete=InService - This test checks if the Notebook stops before deleting.
- Delete=Deleting - This test checks if the Notebook requeues while deleting
- Delete=Pending - This test checks if the Notebook requeues while in pending state
- Delete=Stopping - This test checks if the Notebook requeues while stopping
- Delete=Updating - This test checks if the Notebook requeues while updating.
- Delete=Stopped - This test checks if the Notebook requeues right after making the delete call.
- Delete=Fail - This test checks if the Notebook returns an error if delete fails.
- Delete=StopFail - This test checks if the Notebook returns an error if delete fails.

Nomenclature:
For readone/delete/update output
nb_(operation)__(suffix)
For update input:
nb_update_<desired/latest>_(suffix)

Directory structure:
testdata
  - v1alpha - (folder which has yaml files)
    -  crud operation
       - desired (folder which has input yaml files)
       - observed (folder which has output yaml files)
  - sdkapi
     - readOne - (server responses for readone)
     - create - (server responses for create) 

TODO: 

ReadOne
Replace status with annotation in the following files:
  - testdata/v1alpha1/readone/desired/nb_readone_stopped_after_update_ack.yaml
  - testdata/v1alpha1/readone/desired/nb_readone_stopped_before_update_ack.yaml
  - testdata/v1alpha1/readone/desired/nb_readone_stopping_update.yaml
  - testdata/v1alpha1/readone/observed/nb_readone_stopped_before_update_ack.yaml
  - testdata/v1alpha1/readone/observed/nb_readone_stopped_before_update_ack.yaml
  - testdata/v1alpha1/readone/observed/nb_readone_stopping_update.yaml

Update tests:
Replace status with annotation in the following files:
  - testdata/v1alpha1/update/desired/nb_update_latest_stopped_controller_input.yaml
  - testdata/v1alpha1/update/desired/nb_update_latest_stopped_dissassociated.yaml
  - testdata/v1alpha1/update/desired/nb_update_latest_stopping_input.yaml
  - testdata/v1alpha1/update/observed/nb_update_inservice.yaml
  - testdata/v1alpha1/update/observed/nb_update_stopped_controller.yaml

Delete tests:
Add tests for complete deletion(once unit test supports that)

### CODE REVIEW

File: pkg/resource/notebook_instance/testdata/test_suite.yaml [hunk 1]
```diff
@@ -0,0 +1,319 @@
...
+        given:
+          desired_state: "notebook_instance/v1alpha1/input/update/nb_update_desired_common.yaml"
+          latest_state: "notebook_instance/v1alpha1/input/update/nb_update_latest_stopped_dissassociated.yaml"
+          svc_api:
+            - operation: UpdateNotebookInstanceWithContext
+        invoke: Update
+        expect:
+          latest_state: "notebook_instance/v1alpha1/output/nb_update_dissassociated.yaml"
+
+  #TODO: Add tests for Stop errors once unit test supports that. Also replace status with annotation in the following files:
```
Comment (surajkota) [inline pkg/resource/notebook_instance/testdata/test_suite.yaml:249]: what changes are needed for this? framework already supports throwing error from APIs

cant you do something similar to the `Create=InvalidInput` test?

Comment (ananth102) [reply pkg/resource/notebook_instance/testdata/test_suite.yaml:249]: I'm not exactly sure, for some reason I cant get an error to return for Start/StopNotebookInstance. I inserted a print statement here(https://github.com/aws-controllers-k8s/sagemaker-controller/blob/cd9c3be74768a64f25fb67e796f83a481643ca16/pkg/resource/notebook_instance/custom_sdk_api.go#L28) that printed the error and it printed nil on the test where StartNotebookInstance was supposed to return an error.

Comment (ananth102) [reply pkg/resource/notebook_instance/testdata/test_suite.yaml]: Nevermind it seems to work, might have had a typo when I tried it before.

File: pkg/resource/notebook_instance/testdata/test_suite.yaml [hunk 2]
```diff
@@ -0,0 +1,358 @@
...
+           svc_api:
+             - operation: DescribeNotebookInstanceWithContext
+               error:
+                code: ServiceUnavailable
+                message: "Server is down"
+         invoke: ReadOne
+         expect:
+           error: "ServiceUnavailable: Server is down\n\tstatus code: 0, request id: "
+       - name: "ReadOne=FailedState"
+         description: "Testing how readone handles the Failed State, it should have a failure reason."
```
Comment (ryansteakley) [inline pkg/resource/notebook_instance/testdata/test_suite.yaml:64]: nit: inconsistent naming of `readone`. In most other descriptions you use `readOne`.

File: pkg/resource/notebook_instance/testdata/test_suite.yaml [hunk 3]
```diff
@@ -0,0 +1,358 @@
...
+         given: 
+           desired_state: "notebook_instance/v1alpha1/input/readone/nb_readone_failed_state.yaml"
+           svc_api:
+             - operation: DescribeNotebookInstanceWithContext
+               output_fixture: "notebook_instance/readOne/readone_failed_state.json"
+         invoke: ReadOne
+         expect:
+           latest_state: "notebook_instance/v1alpha1/output/readone/nb_readone_failed_state.yaml"
+       - name: "ReadOne=InService"
+         description: "Testing how readone handles the InService state. Resource Synced should be true"
```
Comment (ryansteakley) [inline pkg/resource/notebook_instance/testdata/test_suite.yaml:74]: same as above

File: pkg/resource/notebook_instance/testdata/test_suite.yaml [hunk 4]
```diff
@@ -0,0 +1,358 @@
...
+         given: 
+           desired_state: "notebook_instance/v1alpha1/input/readone/nb_readone_inservice.yaml"
+           svc_api:
+             - operation: DescribeNotebookInstanceWithContext
+               output_fixture: "notebook_instance/readOne/readone_inservice.json"
+         invoke: ReadOne
+         expect:
+           latest_state: "notebook_instance/v1alpha1/output/readone/nb_readone_inservice.yaml"
+       - name: "ReadOne=StoppedAfterUpdateACK"
+         description: "Testing how readone handles the Stopped after it's been updated. The Notebook here has been stopped by ack."
```
Comment (ryansteakley) [inline pkg/resource/notebook_instance/testdata/test_suite.yaml:84]: same as above

File: pkg/resource/notebook_instance/testdata/test_suite.yaml [hunk 5]
```diff
@@ -0,0 +1,358 @@
...
+         given: 
+           desired_state: "notebook_instance/v1alpha1/input/readone/nb_readone_stopping_deleting.yaml"
+           svc_api:
+             - operation: DescribeNotebookInstanceWithContext
+               output_fixture: "notebook_instance/readOne/readone_stopping_deleting.json"
+         invoke: ReadOne
+         expect:
+           latest_state: "notebook_instance/v1alpha1/output/readone/nb_readone_stopping_deleting.yaml"
+       - name: "ReadOne=StoppedUpdate"
+         description: "Testing how readone handles the Stopping stopping state before going into Update"
```
Comment (ryansteakley) [inline pkg/resource/notebook_instance/testdata/test_suite.yaml:127]: nit: Stopping stopping? repeated word or is this intendend

File: pkg/resource/notebook_instance/testdata/test_suite.yaml [hunk 6]
```diff
@@ -0,0 +1,358 @@
...
+         given: 
+           desired_state: "notebook_instance/v1alpha1/input/readone/nb_readone_stopping_update.yaml"
+           svc_api:
+             - operation: DescribeNotebookInstanceWithContext
+               output_fixture: "notebook_instance/readOne/readone_stopping_update.json"
+         invoke: ReadOne
+         expect:
+           latest_state: "notebook_instance/v1alpha1/output/readone/nb_readone_stopping_update.yaml"
+       - name: "ReadOne=UpdatingUser"
+         description: "Testing how readone handles the Updating status when a user stops the Notebook. StoppedByControllerMeta should
```
Comment (ryansteakley) [inline pkg/resource/notebook_instance/testdata/test_suite.yaml:137]: where is there? StoppedByControllerMeta should not be in status?

File: pkg/resource/notebook_instance/testdata/test_suite.yaml [hunk 7]
```diff
@@ -0,0 +1,358 @@
...
+         given: 
+           desired_state: "notebook_instance/v1alpha1/input/readone/nb_readone_updating_by_user.yaml"
+           svc_api:
+             - operation: DescribeNotebookInstanceWithContext
+               output_fixture: "notebook_instance/readOne/readone_updating_by_user.json"
+         invoke: ReadOne
+         expect:
+           latest_state: "notebook_instance/v1alpha1/output/readone/nb_readone_updating_by_user.yaml"
+       - name: "ReadOne=UpdatingUser"
+         description: "Testing how readone handles the Updating status when ack stops the Notebook"
```
Comment (ryansteakley) [inline pkg/resource/notebook_instance/testdata/test_suite.yaml:148]: nit: ack should be ACK

File: pkg/resource/notebook_instance/testdata/test_suite.yaml [hunk 8]
```diff
@@ -0,0 +1,358 @@
...
+        description: "This test checks if the Notebook requeues while in updating state."
+        given:
+          desired_state: "notebook_instance/v1alpha1/input/update/nb_update_desired_common.yaml"
+          latest_state: "notebook_instance/v1alpha1/input/update/nb_update_latest_updating_input.yaml"
+        invoke: Update
+        expect:
+          error: "NotebookInstance in Updating state cannot be modified or deleted."
+      - name: "Update=InService"
+        description: "This test checks if the controller stops the Notebook. Note: Although a stop controller call has been made
+        runtime has not performed an sdk.Find as a result the NotebookInstanceStatus will be InService."
```
Comment (ryansteakley) [inline pkg/resource/notebook_instance/testdata/test_suite.yaml:209]: nit: sdk.Find()

File: pkg/resource/notebook_instance/testdata/test_suite.yaml (hunk 5 above)
Comment (ananth102) [reply pkg/resource/notebook_instance/testdata/test_suite.yaml:127]: It was a repeated word,  also fixed the test name

File: pkg/resource/notebook_instance/testdata/test_suite.yaml (hunk 6 above)
Comment (ananth102) [reply pkg/resource/notebook_instance/testdata/test_suite.yaml:137]: edited it

File: pkg/resource/notebook_instance/testdata/test_suite.yaml (hunk 2 above)
Comment (ananth102) [reply pkg/resource/notebook_instance/testdata/test_suite.yaml:64]: Going to use readone from now on just to be consistent.

**Review by ryansteakley — APPROVED** [review]

**Comment by surajkota** [comment]
Nice work on the directory structure. What do you think about the following:

for API responses:

```
testdata/sdkapi/create
testdata/sdkapi/describe
testdata/sdkapi/update
```

for test files

for desired
```
testdata/v1alpha1/desired/create
testdata/v1alpha1/desired/read
```

for output i.e. created/observed/updated
```
testdata/v1alpha1/observed/create
testdata/v1alpha1/observed/read
```

**Comment by ananth102** [comment]
@surajkota Yea that sounds better, having a directory for sdkapi also makes it more organized.

**Review by akartsky — APPROVED** [review]

**Pull request merged**